src/ophys_etl/modules/segmentation/processing_log.py

```diff
@@ -0,0 +1,320 @@
+                del f[self.steps_dataset_name]
+            # get unique group name, if necessary. e.g. multiple
+            # steps called 'filter'
+            increment = 0
+            for step in steps:
+                if step.startswith(group_name):
+                    increment += 1
+            if increment != 0:
+                logger.warning(f"{group_name} already exists in log")
+                group_name = f"{group_name}_{increment}"
+                logger.warning(f"creating new group {group_name}")
+            steps.append(group_name)
+            f.create_dataset(self.steps_dataset_name,
+                             data=[i.encode("utf-8") for i in steps])
+        return group_name
+
+    def get_last_group(self) -> str:
+        with h5py.File(self.path, "r") as f:
+            steps = [i.decode("utf-8")
+                     for i in f[self.steps_dataset_name][()]]
+        return steps[-1]
+
+    def log_detection(self,
+                      attribute: str,
+                      rois: List[ExtractROI],
+                      group_name: str = "detect") -> None:
+        """log the detection phase of segmentation to a file
+
+        Parameters
+        ----------
+        attribute: str
+            the name of the edge attribute extracted from the
+            input graph
+        rois: List[ExtractROI]
+            the list of ROIs resulting from detection
+        group_name: str
+            the name of the hdf5 group for logging this step
+            (default = 'detect')
+
+        """
+        group_name = self.append_processing_step(group_name, overwrite=True)
+        with h5py.File(self.path, "a") as h5file:
+            if group_name in h5file:
+                # only one detect step per QC file
+                del h5file[group_name]
+                logging.warn(f"overwriting existing {group_name} group in "
+                             f"{self.path}")
+            group = h5file.create_group(group_name)
+            group.create_dataset("attribute", data=attribute)
+            group.create_dataset("rois",
+                                 data=roi_utils.serialize_extract_roi_list(
+                                     rois))
+            timestamp_group(group)
+
+    def log_merge(self,
+                  rois: List[ExtractROI],
+                  merger_ids: np.ndarray,
+                  group_name: str = "merge") -> None:
+        """log the merge phase of segmentation to a file
+
+        Parameters
+        ----------
+        rois: List[ExtractROI]
+            the list of ROIs resulting from detection
+        group_name: str
+            the name of the hdf5 group for logging this step
+            (default = 'detect')
+        merger_ids: np.ndarray
+            an n_merge x 2 shape array, each row indicating (dst, src)
+            ROI IDs from a merge
+
+        """
+        group_name = self.append_processing_step(group_name, overwrite=False)
```

Comment:
I ultimately do not think we should allow multiple chains of operations to be stored in the same HDF5 file. I just want to call this out in case you disagree on that point.

Reply:
I agree with you. I will implement the post_load concept.

Reply:
done.